### PR TITLE
Set visibility of players/hide them from others

### PR DESCRIPTION
**Pull request by Lejo1**
Rebase of #8397

When players visibility is = false he's not shown as object, in get_player_names or in the /status info

### CODE REVIEW

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -781,6 +781,45 @@ int ObjectRef::l_get_properties(lua_State *L)
...
+	if (!player)
+		return 0;
+
+	bool visible = readParam<bool>(L, 2);
+	co->setVisible(visible);
+
+	getServer(L)->sendFakeJoinLeaveMessage(player->getName(), visible);
+	lua_pushboolean(L, true);
+
+	return 1;
```
Comment (rubenwardy) [inline src/script/lua_api/l_object.cpp:805]: useless and undocumented return value

Comment (Lejo1) [reply src/script/lua_api/l_object.cpp:804]: Removed

File: src/script/lua_api/l_object.cpp [hunk 2]
```diff
@@ -781,6 +781,44 @@ int ObjectRef::l_get_properties(lua_State *L)
...
+	ObjectRef *ref = checkobject(L, 1);
+	ServerActiveObject *co = getobject(ref);
+
+	if (!co)
+		return 0;
+
+	lua_pushboolean(L, co->getVisible());
+
+	return 1;
+}
```
Comment (Zughy) [inline src/script/lua_api/l_object.cpp:820]: Same as before
```
        NO_MAP_LOCK_REQUIRED;
	ObjectRef *ref = checkobject(L, 1);
	ServerActiveObject *sao = getobject(ref);
	if (sao == nullptr)
		return 0;

	lua_pushboolean(L, sao->getVisible());
	return 1;
```
(or `!sao`)

**Review by Zughy — CHANGES_REQUESTED** [review]
Really nice (and needed) feature, bumpbump

File: src/script/lua_api/l_object.cpp [hunk 3]
```diff
@@ -781,6 +781,44 @@ int ObjectRef::l_get_properties(lua_State *L)
...
+
+	if (!player)
+		return 0;
+
+	bool visible = readParam<bool>(L, 2);
+	co->setVisible(visible);
+
+	getServer(L)->sendFakeJoinLeaveMessage(player->getName(), visible);
+
+	return 1;
```
[the remark this answers is not in the document]
Comment (Lejo1) [reply src/script/lua_api/l_object.cpp:804]: Thanks, will just do that than!

File: src/clientiface.cpp [hunk 4]
```diff
@@ -803,6 +807,26 @@ ClientState ClientInterface::getClientState(session_t peer_id)
 	return n->second->getState();
 }
 
+void ClientInterface::sendFakeJoinLeaveMessage(const std::string &player_name,
+		bool in_game)
+{
+	if (in_game) {
```
Comment (rubenwardy) [inline src/clientiface.cpp:813]:
```c++
NetworkPacket notice(TOCLIENT_UPDATE_PLAYER_LIST, 0,
	PEER_ID_INEXISTENT);
notice << (u8) (in_game ? PLAYER_LIST_ADD : PLAYER_LIST_REMOVE)
		<< (u16) 1 << player_name;
sendToAll(&notice);

UpdatePlayerList();
```

**Review by rubenwardy — COMMENTED** [review]
There's already an object property called [`is_visible`](https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L6829), this would be confusing to modders. 

This could be changed to use that object property, except that:

* is_visible=false objects are still collidable
* children of invisible objects may be visible

Perhaps this should be renamed to `set_hidden()` or `set_enabled()`. Android has a ternary for this - `visible`, `invisible`, `gone`.

File: src/clientiface.cpp (hunk 4 above)
Comment (Lejo1) [reply src/clientiface.cpp]: Thanks, added!

File: src/script/lua_api/l_object.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lejo1) [reply src/script/lua_api/l_object.cpp]: changed.

File: src/script/lua_api/l_object.cpp (hunk 2 above)
Comment (Lejo1) [reply src/script/lua_api/l_object.cpp]: changed.

**Comment by Lejo1** [comment]
Used set_player_hidden as it's pretty different to the object is_visible one.

File: src/script/lua_api/l_object.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zughy) [reply src/script/lua_api/l_object.cpp]: I meant empty lines too. To follow the order `check` -> `variables declaration` -> `execution+return`

[the remark this answers is not in the document]
Comment (Lejo1) [reply src/script/lua_api/l_object.cpp]: oh, right changed again

**Comment by Lejo1** [comment]
Rebased

**Comment by AFCMS** [comment]
@Lejo1 Can the player still be seen by using hack clients? In other words, can this pr fully solve #11693 ?

**Comment by Lejo1** [comment]
It fully hides the client even by hack clients, but also as you said fakes the get_player_names entry.

**Comment by Zughy** [comment]
@Lejo1 it should be rebased again, sorry..

**Comment by Lejo1** [comment]
Rebased.

**Comment by SmallJoker** [comment]
https://irc.minetest.net/minetest-dev/2022-09-25#i_6018734

 A few notes, mainly inputs from my side:

 * The PR does too much. Let mods implement chat messages (status, join/leave) for hidden players, if that's wanted
 * Hidden player's attachments must not be sent to other players or they likely stack at (0,0,0) due to unknown parent.
 * What about attached sounds / particlespawners ?

Quick & dirty implementation idea without needing this PR:

 * Set the player's texture to a transparent pixel (blank.png ?)
 * `pointable = false` object property

My implementation idea:

 * Extend `is_visible` on players to hide the player mesh
 * Hide  attachments for `is_visible` parents (fake removal & re-add could get rather messy)
 * Let mods deal with the rest

**Pull request closed without merging**